packages/embeds/embed-core/src/embed-iframe.ts

```diff
@@ -40,32 +47,86 @@ declare global {
       applyCssVars: (cssVarsPerTheme: UiConfig["cssVarsPerTheme"]) => void;
     };
     // Marks that Booker has moved to some non-"loading" state
-    _embedBookerState?: "initializing" | "done";
+    _embedBookerState?: EmbedBookerState;
   }
 }
 
 /**
  * This is in-memory persistence needed so that when user browses through the embed, the configurations from the instructions aren't lost.
  */
-const embedStore = {
+export const embedStore = {
+  /**
+   * Tracks whether the prerender has been completed or not.
+   * prerenderState would be "completed" even after the iframe was switched from isPrerendering to notPrerendering(which happensafter connect)
+   */
+  prerenderState: null as null | "inProgress" | "completed",
   // Handles the commands of routing received from parent even when React hasn't initialized and nextRouter isn't available
   router: {
-    setNextRouter(nextRouter: ReturnType<typeof useRouter>) {
-      this.nextRouter = nextRouter;
+    /**
+     * When we do the history push, it is possible that
+     * - React might revert that change depending on in what state React is in while initializing
+     * - So, we use a declarative approach to ensure that our requirement is continuously met
+     */
+    ensureQueryParamsInUrl({
```

Comment:
Instead of just pushing to next.js router, which could fail depending on if React is already rendering. We are now continuously replacing if the URL has params not according to our needs.

- Next.js App router automatically ensures that on history.push/replace state is updated. 
-Recursive requestAnimationFrame ensures that even if react reverts it, we revert it back to what we need and continue doing that.